File: app/src/main/java/com/usblanbridge/core/EscPosResponder.java
package com.usblanbridge.core;

import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;

/**
 * Answers the status and identity queries POS software sends down the raw 9100 stream, on behalf of a USB
 * printer that cannot reply through the phone's USB pipe in the way clients expect.
 *
 * Every value here was captured from a real Epson TM-T20II, so SDK clients that verify a printer before
 * accepting it behave the same against the bridge as against genuine hardware:
 *
 *   DLE EOT 1 -> 0x16     DLE EOT 2..4 -> 0x12
 *   GS  I  1  -> 0x63     GS  I  67    -> 0x5F "TM-T20II" 0x00
 *   GS  a  n  -> 14 00 00 0F
 *   GS  ( H   -> 0x37 0x22 id 0x00      the process-id echo the Epson SDK waits for to confirm a job finished
 *
 * Queries are consumed rather than forwarded, because they are not print data. Everything else passes through
 * untouched, including print commands sharing the same lead byte such as GS ( k for QR codes.
 *
 * This is a port of the C# EscPosResponder in the Windows build; keep the two in step.
 */
public final class EscPosResponder {

    public static final byte DLE = 0x10;
    public static final byte EOT = 0x04;
    public static final byte GS = 0x1D;

    private static final int MAX_HELD = 64 * 1024;

    private String modelName = "TM-T20II";

    private byte[] pending = new byte[256];
    private int pendingLen;

    private final ByteArrayOutputStream forward = new ByteArrayOutputStream(8192);
    private final ByteArrayOutputStream replies = new ByteArrayOutputStream(64);

    private int queriesAnswered;

    public void setModelName(String name) {
        if (name != null && name.length() > 0) this.modelName = name;
    }

    public int getQueriesAnswered() {
        return queriesAnswered;
    }

    /** Bytes to send on to the printer after the last {@link #process} call. */
    public byte[] forwardBytes() {
        return forward.toByteArray();
    }

    public int forwardLength() {
        return forward.size();
    }

    /** Bytes to send back to the network client after the last {@link #process} call. */
    public byte[] replyBytes() {
        return replies.toByteArray();
    }

    public int replyLength() {
        return replies.size();
    }

    private enum Match {COMPLETE, INCOMPLETE, NO}

    public void process(byte[] input, int offset, int count) {
        forward.reset();
        replies.reset();

        ensurePending(pendingLen + count);
        System.arraycopy(input, offset, pending, pendingLen, count);
        pendingLen += count;

        int pos = 0;
        while (pos < pendingLen) {
            byte b = pending[pos];
            if (b != DLE && b != GS) {
                forward.write(b);
                pos++;
                continue;
            }
            int[] consumed = new int[1];
            Match m = tryMatch(pos, consumed);
            if (m == Match.COMPLETE) {
                pos += consumed[0];
                queriesAnswered++;
            } else if (m == Match.NO) {
                forward.write(b);
                pos++;
            } else {
                break; // incomplete: wait for the rest of the query
            }
        }

        if (pos > 0) {
            System.arraycopy(pending, pos, pending, 0, pendingLen - pos);
            pendingLen -= pos;
        }

        // Never hold an unbounded amount of data waiting for a query that will never complete.
        if (pendingLen > MAX_HELD) flushPending();
    }

    /** Releases bytes held back as a possible query. Call when the client disconnects or a job ends. */
    public void flush() {
        forward.reset();
        replies.reset();
        flushPending();
    }

    private void flushPending() {
        forward.write(pending, 0, pendingLen);
        pendingLen = 0;
    }

    private int available(int from) {
        return pendingLen - from;
    }

    private Match tryMatch(int pos, int[] consumed) {
        consumed[0] = 0;
        byte lead = pending[pos];

        if (lead == DLE) {
            if (available(pos) < 2) return Match.INCOMPLETE;
            if (pending[pos + 1] != EOT) return Match.NO; // DLE DC4 drawer pulse and friends pass through
            if (available(pos) < 3) return Match.INCOMPLETE;
            int n = pending[pos + 2] & 0xFF;
            if (n < 1 || n > 4) return Match.NO;
            replies.write(n == 1 ? 0x16 : 0x12);
            consumed[0] = 3;
            return Match.COMPLETE;
        }

        if (available(pos) < 2) return Match.INCOMPLETE;
        int second = pending[pos + 1] & 0xFF;

        if (second == 0x49) { // GS I n
            if (available(pos) < 3) return Match.INCOMPLETE;
            replyPrinterId(pending[pos + 2] & 0xFF);
            consumed[0] = 3;
            return Match.COMPLETE;
        }

        if (second == 0x61) { // GS a n, enable automatic status back
            if (available(pos) < 3) return Match.INCOMPLETE;
            replies.write(0x14);
            replies.write(0x00);
            replies.write(0x00);
            replies.write(0x0F);
            consumed[0] = 3;
            return Match.COMPLETE;
        }

        if (second == 0x72) { // GS r n, transmit status
            if (available(pos) < 3) return Match.INCOMPLETE;
            replies.write(0x00);
            consumed[0] = 3;
            return Match.COMPLETE;
        }

        if (second == 0x28) { // GS ( ... ; only GS ( H is a query
            if (available(pos) < 3) return Match.INCOMPLETE;
            if ((pending[pos + 2] & 0xFF) != 0x48) return Match.NO;
            if (available(pos) < 5) return Match.INCOMPLETE;
            int len = (pending[pos + 3] & 0xFF) | ((pending[pos + 4] & 0xFF) << 8);
            if (len < 2 || len > 1024) return Match.NO; // implausible: treat as print data
            int total = 5 + len;
            if (available(pos) < total) return Match.INCOMPLETE;

            replies.write(0x37);
            replies.write(0x22);
            for (int i = pos + 7; i < pos + total; i++) replies.write(pending[i]);
            replies.write(0x00);
            consumed[0] = total;
            return Match.COMPLETE;
        }

        return Match.NO;
    }

    private void replyPrinterId(int n) {
        switch (n) {
            case 1:
            case 49:
                replies.write(0x63);
                break;
            case 2:
            case 50:
            case 3:
            case 51:
                replies.write(0x00);
                break;
            default:
                if (n >= 65) {
                    replies.write(0x5F);
                    byte[] name;
                    try {
                        name = modelName.getBytes("US-ASCII");
                    } catch (UnsupportedEncodingException e) {
                        name = modelName.getBytes();
                    }
                    replies.write(name, 0, name.length);
                    replies.write(0x00);
                } else {
                    replies.write(0x00);
                }
                break;
        }
    }

    private void ensurePending(int needed) {
        if (pending.length >= needed) return;
        int size = pending.length;
        while (size < needed) size *= 2;
        byte[] bigger = new byte[size];
        System.arraycopy(pending, 0, bigger, 0, pendingLen);
        pending = bigger;
    }
}

File: src/UsbLanPrinterBridge/Core/EscPosResponder.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace UsbLanPrinterBridge.Core
{
    /// <summary>
    /// Answers the status and identity queries that Epson's ePOS SDK (and most POS software) sends down the raw
    /// 9100 stream, on behalf of a USB printer that can never reply through the Windows spooler.
    ///
    /// Every response value here was captured from a real Epson TM-T20II on the LAN, so clients that verify the
    /// printer before accepting it behave the same against the bridge as against genuine hardware:
    ///
    ///   DLE EOT 1  -> 0x16              printer status, online, no error
    ///   DLE EOT 2  -> 0x12              offline cause, none
    ///   DLE EOT 3  -> 0x12              error cause, none
    ///   DLE EOT 4  -> 0x12              paper sensor, paper present
    ///   GS  I  1   -> 0x63              printer model id
    ///   GS  I  67  -> 0x5F "TM-T20II" 0x00   model name, header + text + NUL
    ///   GS  a  n   -> 14 00 00 0F       automatic status back, 4 bytes
    ///   GS  ( H    -> 0x37 0x22 id 0x00 process id echo: how the SDK learns a job finished
    ///
    /// The queries are consumed rather than forwarded: they are not print data, and passing them to the spooler
    /// would create empty jobs. Everything else passes through untouched, including print commands that start with
    /// the same lead bytes such as GS ( k for QR codes.
    ///
    /// The parser is incremental. A query split across TCP reads is still recognised, and at most a few bytes are
    /// held back between calls; <see cref="Flush"/> releases them.
    /// </summary>
    public sealed class EscPosResponder
    {
        public const byte DLE = 0x10;
        public const byte EOT = 0x04;
        public const byte GS = 0x1D;

        /// <summary>Model name reported to GS I 67. Mirrors a widely supported Epson model so SDK clients accept it.</summary>
        public string ModelName { get; set; }

        private readonly List<byte> _buffer = new List<byte>(64);
        private byte[] _forward = new byte[8192];
        private byte[] _replies = new byte[64];

        public EscPosResponder() { ModelName = "TM-T20II"; }

        public byte[] Forward { get { return _forward; } }
        public int ForwardLength { get; private set; }
        public byte[] Replies { get { return _replies; } }
        public int ReplyLength { get; private set; }

        /// <summary>Number of queries answered so far (diagnostics).</summary>
        public int QueriesAnswered { get; private set; }

        /// <summary>The last few queries answered, by name, oldest first (diagnostics for the printer actions log).</summary>
        public List<string> RecentQueries { get { return _recent; } }
        private readonly List<string> _recent = new List<string>();
        private const int RecentLimit = 12;

        private void Note(string query)
        {
            if (_recent.Count >= RecentLimit) _recent.RemoveAt(0);
            _recent.Add(query);
        }

        private enum Match { Complete, Incomplete, No }

        public void Process(byte[] input, int offset, int count)
        {
            ForwardLength = 0;
            ReplyLength = 0;
            EnsureCapacity(ref _forward, count + _buffer.Count + 8);

            for (int i = 0; i < count; i++) _buffer.Add(input[offset + i]);

            int pos = 0;
            while (pos < _buffer.Count)
            {
                byte b = _buffer[pos];
                if (b != DLE && b != GS) { Emit(b); pos++; continue; }

                int consumed;
                Match m = TryMatch(pos, out consumed);
                if (m == Match.Complete) { pos += consumed; QueriesAnswered++; }
                else if (m == Match.No) { Emit(b); pos++; }
                else break; // incomplete: keep the rest for the next chunk
            }

            if (pos > 0) _buffer.RemoveRange(0, pos);

            // Safety valve: never hold an unbounded amount of data waiting for a query that will not complete.
            if (_buffer.Count > 64 * 1024) FlushBuffer();
        }

        /// <summary>Releases bytes held back as a possible query. Call when the client disconnects or the job ends.</summary>
        public void Flush()
        {
            ForwardLength = 0;
            ReplyLength = 0;
            FlushBuffer();
        }

        private void FlushBuffer()
        {
            EnsureCapacity(ref _forward, ForwardLength + _buffer.Count);
            for (int i = 0; i < _buffer.Count; i++) Emit(_buffer[i]);
            _buffer.Clear();
        }

        private byte At(int index) { return _buffer[index]; }
        private int Available(int from) { return _buffer.Count - from; }

        private Match TryMatch(int pos, out int consumed)
        {
            consumed = 0;
            byte lead = At(pos);

            if (lead == DLE)
            {
                // DLE EOT n : real-time status request.
                if (Available(pos) < 2) return Match.Incomplete;
                if (At(pos + 1) != EOT) return Match.No;      // e.g. DLE DC4 drawer pulse: pass through
                if (Available(pos) < 3) return Match.Incomplete;
                byte n = At(pos + 2);
                if (n < 1 || n > 4) return Match.No;
                Reply(StatusFor(n));
                Note("DLE EOT " + n + " (" + (n == 1 ? "printer" : n == 2 ? "offline" : n == 3 ? "error" : "paper") + " status)");
                consumed = 3;
                return Match.Complete;
            }

            // lead == GS
            if (Available(pos) < 2) return Match.Incomplete;
            byte second = At(pos + 1);

            if (second == 0x49) // GS I n : transmit printer id
            {
                if (Available(pos) < 3) return Match.Incomplete;
                ReplyPrinterId(At(pos + 2));
                Note("GS I " + At(pos + 2) + " (printer id" + (At(pos + 2) >= 65 ? ", model name" : "") + ")");
                consumed = 3;
                return Match.Complete;
            }

            if (second == 0x61) // GS a n : enable automatic status back
            {
                if (Available(pos) < 3) return Match.Incomplete;
                Reply(0x14); Reply(0x00); Reply(0x00); Reply(0x0F);
                Note("GS a (auto status back)");
                consumed = 3;
                return Match.Complete;
            }

            if (second == 0x72) // GS r n : transmit status
            {
                if (Available(pos) < 3) return Match.Incomplete;
                Reply(0x00); // paper present / drawer low
                Note("GS r " + At(pos + 2) + " (paper/drawer status)");
                consumed = 3;
                return Match.Complete;
            }

            if (second == 0x28) // GS ( ... : only GS ( H is a query; GS ( k and friends are print data
            {
                if (Available(pos) < 3) return Match.Incomplete;
                if (At(pos + 2) != 0x48) return Match.No;
                if (Available(pos) < 5) return Match.Incomplete;
                int len = At(pos + 3) | (At(pos + 4) << 8);
                if (len < 2 || len > 1024) return Match.No; // implausible: treat as print data
                int total = 5 + len;
                if (Available(pos) < total) return Match.Incomplete;

                // GS ( H pL pH fn m d1..dk  ->  0x37 0x22 d1..dk 0x00
                Reply(0x37);
                Reply(0x22);
                for (int i = pos + 7; i < pos + total; i++) Reply(At(i));
                Reply(0x00);
                Note("GS ( H (process id echo)");
                consumed = total;
                return Match.Complete;
            }

            return Match.No;
        }

        private static byte StatusFor(byte n)
        {
            // Captured from a real TM-T20II: n=1 -> 0x16, n=2..4 -> 0x12.
            return n == 1 ? (byte)0x16 : (byte)0x12;
        }

        private void ReplyPrinterId(byte n)
        {
            switch (n)
            {
                case 1:
                case 49:
                    Reply(0x63); // printer model id
                    break;
                case 2:
                case 50:
                    Reply(0x00); // type id
                    break;
                case 3:
                case 51:
                    Reply(0x00); // firmware version id
                    break;
                default:
                    if (n >= 65)
                    {
                        // Header + text + NUL, the format the real printer uses for GS I 67.
                        Reply(0x5F);
                        foreach (byte c in Encoding.ASCII.GetBytes(ModelName ?? "TM-T20II")) Reply(c);
                        Reply(0x00);
                    }
                    else Reply(0x00);
                    break;
            }
        }

        private void Emit(byte b)
        {
            EnsureCapacity(ref _forward, ForwardLength + 1);
            _forward[ForwardLength++] = b;
        }

        private void Reply(byte b)
        {
            EnsureCapacity(ref _replies, ReplyLength + 1);
            _replies[ReplyLength++] = b;
        }

        private static void EnsureCapacity(ref byte[] buffer, int needed)
        {
            if (buffer.Length >= needed) return;
            int size = buffer.Length;
            while (size < needed) size *= 2;
            Array.Resize(ref buffer, size);
        }
    }
}
